daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala

```diff
@@ -375,188 +374,245 @@ object ScenarioLedger {
+    * @param trId transaction identity
+    * @param richTr (enriched) transaction
+    * @param locationInfo location map
+    */
+  class TransactionProcessor(
       trId: TransactionId,
       richTr: RichTransaction,
       locationInfo: Map[NodeId, Location],
-      ledgerData: LedgerData,
-  ): Either[UniqueKeyViolation, LedgerData] = {
+  ) {
 
-    final case class RollbackBeginState(
-        activeContracts: Set[ContractId],
-        activeKeys: Map[GlobalKey, ContractId],
-    )
+    def duplicateKeyCheck(ledgerData: LedgerData): Either[UniqueKeyViolation, Unit] = {
+      val inactiveKeys = richTr.transaction.contractKeyInputs
+        .fold(error => crash(s"$error: inconsistent transaction"), identity)
+        .collect { case (key, _: Tx.KeyInactive) =>
+          key
+        }
 
-    final case class ProcessingNode(
-        // The id of the direct parent or None.
-        mbParentId: Option[NodeId],
-        // The id of the nearest rollback ancestor. If this is
-        // a rollback node itself, it points to itself.
-        mbRollbackAncestorId: Option[NodeId],
-        children: List[NodeId],
-        // For rollback nodes, we store the previous state here and restore it.
-        // For exercise nodes, we don’t need to restore anything.
-        prevState: Option[RollbackBeginState],
-    )
+      inactiveKeys.find(ledgerData.activeKeys.contains(_)) match {
+        case Some(duplicateKey) =>
+          Left(UniqueKeyViolation(duplicateKey))
 
-    @tailrec
-    def processNodes(
-        cache0: LedgerData,
-        enps: List[ProcessingNode],
-    ): LedgerData = enps match {
-      case Nil => cache0
-      case ProcessingNode(_, _, Nil, optPrevState) :: restENPs => {
-        val cache1 = optPrevState.fold(cache0) { case prevState =>
-          cache0.copy(
-            activeContracts = prevState.activeContracts,
-            activeKeys = prevState.activeKeys,
-          )
-        }
-        processNodes(cache1, restENPs)
+        case None =>
+          Right(())
       }
-      case (processingNode @ ProcessingNode(
-            mbParentId,
-            mbRollbackAncestorId,
-            nodeId :: restOfNodeIds,
-            optPrevState,
-          )) :: restENPs =>
-        val eventId = EventId(trId.id, nodeId)
-        richTr.transaction.nodes.get(nodeId) match {
-          case None =>
-            crash(s"processTransaction: non-existent node '$eventId'.")
-          case Some(node) =>
-            val newLedgerNodeInfo = LedgerNodeInfo(
-              node = node,
-              optLocation = locationInfo.get(nodeId),
-              transaction = trId,
-              effectiveAt = richTr.effectiveAt,
-              disclosures = Map.empty,
-              referencedBy = Set.empty,
-              consumedBy = None,
-              rolledbackBy = mbRollbackAncestorId,
-              parent = mbParentId.map(EventId(trId.id, _)),
+    }
+
+    def addNewLedgerNodes(historicalLedgerData: LedgerData): LedgerData =
```

Comment:
you could consider merging createdIn & referencedBy in here. Those don’t care about the tree structure. parentUpdates unfortunately does so merging that one is harder.